question_answering_based_on_context_using_openai_and_pinecone/main.bal

```diff
@@ -0,0 +1,96 @@
+import ballerina/http;
+import ballerina/io;
+import ballerina/regex;
+import ballerinax/googleapis.sheets;
+import ballerinax/openai.embeddings;
+import ballerinax/openai.text;
+import ballerinax/pinecone.vector as pinecone;
+
+configurable string sheetsAccessToken = ?;
+configurable string sheetId = ?;
+configurable string sheetName = ?;
+configurable string openAIToken = ?;
+configurable string pineconeKey = ?;
+configurable string pineconeServiceUrl = ?;
+
+final sheets:Client gSheets = check new ({auth: {token: sheetsAccessToken}});
+final text:Client openAIText = check new ({auth: {token: openAIToken}});
+final embeddings:Client openaiEmbeddings = check new ({auth: {token: openAIToken}});
+final pinecone:Client pineconeClient = check new ({apiKey: pineconeKey}, serviceUrl = pineconeServiceUrl);
+
+const MAXIMUM_NO_OF_DOCS = 10;
+const NAMESPACE = "ChoreoDocs";
+
+service / on new http:Listener(8080) {
+    function init() returns error? {
+        sheets:Range range = check gSheets->getRange(sheetId, sheetName, "A2:B");
+        pinecone:Vector[] vectors = [];
+
+        foreach any[] row in range.values {
+            string title = <string>row[0];
+            string content = <string>row[1];
+            float[] vector = check getEmbedding(string `${title} ${"\n"} ${content}`);
+            vectors[vectors.length()] = {id: title, values: vector, metadata: {"content": content}};
+        }
+
+        pinecone:UpsertResponse response = check pineconeClient->/vectors/upsert.post({vectors, namespace: NAMESPACE});
+        if response.upsertedCount != range.values.length() {
+            return error("Failed to insert embedding vectors to pinecone.");
+        }
+        io:println("Successfully inserted embedding vectors to pinecone.");
+    }
+
+    resource function get answer(string question) returns string?|error {
+        string prompt = check constructPrompt(question);
+        text:CreateCompletionRequest prmt = {
+            prompt: prompt,
+            model: "text-davinci-003",
+            max_tokens: 2000
+        };
+        text:CreateCompletionResponse completionRes = check openAIText->/completions.post(prmt);
+        return completionRes.choices[0].text;
+    }
+}
+
+function getEmbedding(string text) returns float[]|error {
+    embeddings:CreateEmbeddingRequest embeddingRequest = {
+        input: text,
+        model: "text-embedding-ada-002"
+    };
+    embeddings:CreateEmbeddingResponse embeddingRes = check openaiEmbeddings->/embeddings.post(embeddingRequest);
+    return embeddingRes.data[0].embedding;
+}
+
+function countWords(string text) returns int => regex:split(text, " ").length();
+
+function constructPrompt(string question) returns string|error {
+    float[] questionEmbedding = check getEmbedding(question);
+
+    pinecone:QueryRequest req = {namespace: NAMESPACE, topK: MAXIMUM_NO_OF_DOCS, vector: questionEmbedding, includeMetadata: true};
```

Comment:
Break the line appropriately with max char limit of 120.

Reply:
Addressed in https://github.com/ballerina-guides/ai-samples/pull/66/commits/d9c08c77460527472ba049ee200a979dba45dfef.

Reply:
Break as follows.
```ballerina
    pinecone:QueryRequest req = {
        namespace: NAMESPACE,
        topK: MAXIMUM_NO_OF_DOCS,
        vector: questionEmbedding,
        includeMetadata: true
    };
```

Reply:
Addressed in https://github.com/ballerina-guides/ai-samples/pull/66/commits/0cb8e316dac0aafaa633c5883ff1c138b58a53d5.